protocol/msgpack/encoder.go

```diff
@@ -0,0 +1,211 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/m3db/m3aggregator/metric"
+	"github.com/m3db/m3aggregator/policy"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// newBufferedEncoder creates a new buffered encoder
+func newBufferedEncoder() BufferedEncoder {
+	buffer := bytes.NewBuffer(nil)
+
+	return BufferedEncoder{
+		Buffer:  buffer,
+		Encoder: msgpack.NewEncoder(buffer),
+	}
+}
+
+// batchEncoder is a batch encoder that uses MessagePack for encoding.
+// It is NOT thread-safe. User must protect the encoder if accessed from
+// multiple goroutines.
+type batchEncoder struct {
+	opts        EncoderOptions      // encoding options
+	encoderPool BufferedEncoderPool // pool for internal encoders
+	encoder     BufferedEncoder     // internal encoder that does the actual encoding
+	err         error               // error encountered during encoding
+}
+
+// NewBatchEncoder creates a new batch encoder
+func NewBatchEncoder(opts EncoderOptions) (BatchEncoder, error) {
+	if opts == nil {
+		opts = NewEncoderOptions()
+	}
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	encoderPool := opts.BufferedEncoderPool()
+	return &batchEncoder{
+		opts:        opts,
+		encoderPool: encoderPool,
+		encoder:     encoderPool.Get(),
+	}, nil
+}
+
+// StartBatch starts the encoding process
+func (enc *batchEncoder) StartBatch(batchSize int) error {
+	// If the batch size is not positive, this is an invalid request
+	// so we return an error but don't set error in the encoder since
+	// this doesn't corrupt the encoded byte stream
+	if batchSize <= 0 {
+		return fmt.Errorf("batch size must be positive but is instead %d", batchSize)
+	}
+	enc.encodeBatchSize(batchSize)
+	enc.encodeVersion(supportedVersion)
+	return enc.err
+}
+
+func (enc *batchEncoder) EncodeCounter(c metric.Counter, p policy.VersionedPolicies) error {
+	enc.encodeType(metric.CounterType)
```

Comment:
Yeah this might be the c++ way, but unfortunately the go version of the API (https://godoc.org/gopkg.in/vmihailenco/msgpack.v2#Encoder.Encode) converts each argument passed in to an interface{}, and then try to figure out what the type is (sometimes via reflection too), and then encodes accordingly based on the type. That's a big hit on perf based on my benchmark, and adds GC overhead too due to the extra allocations we need to do for the interface conversion, so for the simple types I'm trying to serialize and deserialize, I'm resorting to the lower-level APIs that don't require this sort of interface conversions.